Mathlib/NumberTheory/Padics/PadicVal.lean

```diff
@@ -95,6 +97,37 @@ theorem eq_zero_of_not_dvd {n : ℕ} (h : ¬p ∣ n) : padicValNat p n = 0 :=
   eq_zero_iff.2 <| Or.inr <| Or.inr h
 #align padic_val_nat.eq_zero_of_not_dvd padicValNat.eq_zero_of_not_dvd
 
+open Nat.maxPowDiv
+
+theorem maxPowDiv_eq_multiplicity {p n : ℕ} (hp : 1 < p) (hn : 0 < n) :
+    p.maxPowDiv n = multiplicity p n := by
+  apply multiplicity.unique <| pow_dvd p n
+  intro h
+  apply Nat.not_lt.mpr <| le_of_dvd hp hn h
+  simp
+
+theorem maxPowDiv_eq_multiplicity_get {p n : ℕ} (hp : 1 < p) (hn : 0 < n) (h : Finite p n) :
+    p.maxPowDiv n = (multiplicity p n).get h := by
+  rw [PartENat.get_eq_iff_eq_coe.mpr]
+  apply maxPowDiv_eq_multiplicity hp hn|>.symm
+
+/-- Allows for more efficient code for `padicValNat` -/
+@[csimp]
+theorem padicValNat_eq_maxPowDiv : @padicValNat = @maxPowDiv := by
```

Comment:
Perhaps add a file under `test/` that evaluates `padicValNat`, as a witness that this `@[csimp]` is doing its job? (Only because this is relatively new to be doing this in mathlib, we don't need this for every csimp we introduce once we have confidence.)

Reply:
I've added a test which evaluates `padicValNat 2 (2^100000)`. In core, they are a bit smarter. They [print the generated IR to stdout](https://github.com/leanprover/lean4/blob/bff612e59e4860bef5180db08356ab5f256bea86/tests/lean/csimpAttr.lean#L15) and then check the correct declaration appears. It would nice to do this completely in Lean.